bindings/rust/src/lib.rs

```diff
@@ -289,6 +289,75 @@ impl TypedExecutionResult {
 }
 
 impl Instance {
+    /// Ensure the range is valid according to the currently available memory size.
+    fn checked_memory_range(
+        &self,
+        offset: u32,
+        size: usize,
+    ) -> Result<core::ops::Range<usize>, ()> {
+        // This should be safe given usize::BITS >= u32::BITS, see https://doc.rust-lang.org/std/primitive.usize.html.
+        let offset = offset as usize;
+        let memory_size = self.memory_size();
+        // Empty slices are allowed, but ensure both starting and ending offsets are valid.
+        if memory_size == 0 || offset > memory_size || (offset + size) > memory_size {
```

Comment:
`offset > memory_size` condition is redundant then, when it is true `(offset + size) > memory_size` will always be true.
```suggestion
        if memory_size == 0 || (offset + size) > memory_size {
```

Also for the case of `(memory 0)` it would be perhaps logical to allow (0, 0) slice, then the condition should make an error only the absence of memory
```suggestion
        if !sys::fizzy_module_has_memory(self.get_module()) || (offset + size) > memory_size {
```

Reply:
> offset > memory_size condition is redundant then, when it is true (offset + size) > memory_size will always be true.

Was a bad attempt for overflow check. Decided to just use `offset.checked_add(size).is_none()` for overflow check.